File: workers/fixtures.py
"""Populate the development index with repeatable SideQuest fixture content."""

from __future__ import annotations

import json
import random
from datetime import UTC, datetime, time, timedelta, timezone
from pathlib import Path
from typing import Any

from lib.store import Store


INDEX_NAME = "sidequest-places"
FIXTURE_FILTER = {"source": "fixture", "is_fixture": True}
_LANDMARKS_PATH = Path(__file__).parents[1] / "data" / "landmarks.json"
# The fixture date is the July hackathon date, when San Francisco observes PDT.
# A fixed offset avoids requiring the uninstalled Windows time-zone database.
_SAN_FRANCISCO = timezone(timedelta(hours=-7))

_EVENTS = (
    ("Waterfront sunset walk", 18, "A guided waterfront walk begins near {name}."),
    ("Neighborhood trivia night", 19, "A casual neighborhood trivia night is happening at {name}."),
    ("Bay-view photo meetup", 19, "A photo meetup gathers at {name} for bay views."),
    ("Local makers pop-up", 20, "A small local makers pop-up is on tonight at {name}."),
    ("Evening art conversation", 21, "An informal art conversation is scheduled at {name}."),
)

_CAPTIONS = (
    "Golden-hour colors made this stop worth the walk.",
    "A favorite little city view from today.",
    "Found a calm moment between the downtown rush.",
    "Waterfront light and a breezy afternoon.",
    "Saving this perspective for the next visit.",
    "A quick detour with a surprisingly great backdrop.",
    "City textures, open sky, and a good long walk.",
    "One of those corners that photographs itself.",
    "A pause here made the whole route better.",
    "The view was even better in person.",
    "An easy place to linger with a camera.",
    "Tonight's walk came with a postcard-worthy scene.",
)


def load_fixtures(store: Store, now: datetime | None = None) -> dict[str, int]:
    """Replace the complete fixture feed and return counts written to ``store``."""
    fetched_at = _as_utc(now or datetime.now(UTC))
    landmarks = _load_landmarks()
    busyness_by_place = {
        landmark["place_id"]: random.randint(20, 90) for landmark in landmarks
    }

    # FakeStore has no upsert behavior. Replacing this worker's complete fixture
    # feed before adding records keeps both supported Store backends idempotent.
    store.delete_where(INDEX_NAME, FIXTURE_FILTER)

    landmark_docs = [
        _landmark_doc(landmark, fetched_at, busyness_by_place[landmark["place_id"]])
        for landmark in landmarks
    ]
    store.add_docs(INDEX_NAME, landmark_docs)
    for place_id, busyness_pct in busyness_by_place.items():
        store.update_metadata(
            INDEX_NAME,
            place_id,
            {"busyness_pct": busyness_pct, "busyness_at": fetched_at},
        )

    photogenic = [landmark for landmark in landmarks if landmark["photogenic"]]
    event_docs = _event_docs(photogenic, fetched_at, busyness_by_place)
    caption_docs = _caption_docs(photogenic, fetched_at, busyness_by_place)
    store.add_docs(INDEX_NAME, event_docs + caption_docs)

    return {
        "landmarks": len(landmark_docs),
        "events": len(event_docs),
        "captions": len(caption_docs),
    }


def _load_landmarks() -> list[dict[str, Any]]:
    with _LANDMARKS_PATH.open(encoding="utf-8") as landmarks_file:
        return json.load(landmarks_file)


def _landmark_doc(landmark: dict[str, Any], fetched_at: str, busyness_pct: int) -> dict:
    return {
        "text": f"{landmark['name']} is a SideQuest stop in San Francisco.",
        "metadata": _metadata(landmark, fetched_at, busyness_pct),
    }


def _event_docs(
    photogenic: list[dict[str, Any]], fetched_at: str, busyness_by_place: dict[str, int]
) -> list[dict]:
    tonight = datetime.fromisoformat(fetched_at).astimezone(_SAN_FRANCISCO).date()
    docs = []
    for event_number, (title, hour, description) in enumerate(_EVENTS):
        landmark = photogenic[event_number]
        event_time = datetime.combine(tonight, time(hour), tzinfo=_SAN_FRANCISCO).astimezone(UTC)
        metadata = _metadata(landmark, fetched_at, busyness_by_place[landmark["place_id"]])
        metadata.update({"kind": "event", "event_time": _as_utc(event_time), "heat_score": 6})
        docs.append({"text": f"Event: {title}. {description.format(name=landmark['name'])}", "metadata": metadata})
    return docs


def _caption_docs(
    photogenic: list[dict[str, Any]], fetched_at: str, busyness_by_place: dict[str, int]
) -> list[dict]:
    docs = []
    for caption_number, caption in enumerate(_CAPTIONS):
        landmark = photogenic[caption_number % len(photogenic)]
        metadata = _metadata(landmark, fetched_at, busyness_by_place[landmark["place_id"]])
        metadata.update({"kind": "caption", "heat_score": caption_number % 10 + 1})
        docs.append({"text": f"Instagram-style caption about {landmark['name']}: {caption}", "metadata": metadata})
    return docs


def _metadata(landmark: dict[str, Any], fetched_at: str, busyness_pct: int) -> dict:
    return {
        **landmark,
        "busyness_pct": busyness_pct,
        "busyness_at": fetched_at,
        "heat_score": 1,
        "open_now": True,
        "event_time": None,
        "lang": "en",
        "source": "fixture",
        "fetched_at": fetched_at,
        "is_fixture": True,
    }


def _as_utc(value: datetime) -> str:
    return value.astimezone(UTC).isoformat().replace("+00:00", "Z")

File: scripts/fixtures.ts
/**
 * Populate the development index with repeatable SideQuest fixture content.
 *
 * Ported from workers/fixtures.py.
 */

import { readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

import type { Document, Store } from '../lib/server/store.ts';

export const INDEX_NAME = 'sidequest-places';
export const FIXTURE_FILTER = { source: 'fixture', is_fixture: true };

const LANDMARKS_PATH = join(dirname(fileURLToPath(import.meta.url)), '..', 'data', 'landmarks.json');

/**
 * The fixture date is the July hackathon date, when San Francisco observes PDT.
 * A fixed offset avoids requiring the uninstalled Windows time-zone database.
 */
const SAN_FRANCISCO_UTC_OFFSET_HOURS = -7;

const EVENTS: ReadonlyArray<readonly [string, number, string]> = [
  ['Waterfront sunset walk', 18, 'A guided waterfront walk begins near {name}.'],
  ['Neighborhood trivia night', 19, 'A casual neighborhood trivia night is happening at {name}.'],
  ['Bay-view photo meetup', 19, 'A photo meetup gathers at {name} for bay views.'],
  ['Local makers pop-up', 20, 'A small local makers pop-up is on tonight at {name}.'],
  ['Evening art conversation', 21, 'An informal art conversation is scheduled at {name}.'],
] as const;

const CAPTIONS = [
  'Golden-hour colors made this stop worth the walk.',
  'A favorite little city view from today.',
  'Found a calm moment between the downtown rush.',
  'Waterfront light and a breezy afternoon.',
  'Saving this perspective for the next visit.',
  'A quick detour with a surprisingly great backdrop.',
  'City textures, open sky, and a good long walk.',
  'One of those corners that photographs itself.',
  'A pause here made the whole route better.',
  'The view was even better in person.',
  'An easy place to linger with a camera.',
  "Tonight's walk came with a postcard-worthy scene.",
] as const;

export interface Landmark {
  place_id: string;
  name: string;
  lat: number;
  lng: number;
  photogenic: boolean;
  [key: string]: unknown;
}

export interface FixtureCounts {
  landmarks: number;
  events: number;
  captions: number;
}

/** Replace the complete fixture feed and return counts written to `store`. */
export async function loadFixtures(store: Store, now?: Date): Promise<FixtureCounts> {
  const fetchedAt = asUtc(now ?? new Date());
  const landmarks = await loadLandmarks();

  const busynessByPlace = new Map<string, number>(
    landmarks.map((landmark) => [landmark.place_id, randomInt(20, 90)]),
  );

  // FakeStore has no upsert behavior. Replacing this worker's complete fixture
  // feed before adding records keeps both supported Store backends idempotent.
  await store.deleteWhere(INDEX_NAME, FIXTURE_FILTER);

  const landmarkDocs = landmarks.map((landmark) =>
    landmarkDoc(landmark, fetchedAt, busynessByPlace.get(landmark.place_id)!),
  );
  await store.addDocs(INDEX_NAME, landmarkDocs);

  for (const [placeId, busynessPct] of busynessByPlace) {
    await store.updateMetadata(INDEX_NAME, placeId, {
      busyness_pct: busynessPct,
      busyness_at: fetchedAt,
    });
  }

  const photogenic = landmarks.filter((landmark) => landmark.photogenic);
  const eventDocs = buildEventDocs(photogenic, fetchedAt, busynessByPlace);
  const captionDocs = buildCaptionDocs(photogenic, fetchedAt, busynessByPlace);
  await store.addDocs(INDEX_NAME, [...eventDocs, ...captionDocs]);

  return {
    landmarks: landmarkDocs.length,
    events: eventDocs.length,
    captions: captionDocs.length,
  };
}

async function loadLandmarks(): Promise<Landmark[]> {
  return JSON.parse(await readFile(LANDMARKS_PATH, 'utf-8')) as Landmark[];
}

/** Inclusive on both bounds, matching Python's random.randint. */
function randomInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function landmarkDoc(landmark: Landmark, fetchedAt: string, busynessPct: number): Document {
  return {
    text: `${landmark.name} is a SideQuest stop in San Francisco.`,
    metadata: buildMetadata(landmark, fetchedAt, busynessPct),
  };
}

function buildEventDocs(
  photogenic: Landmark[],
  fetchedAt: string,
  busynessByPlace: Map<string, number>,
): Document[] {
  const docs: Document[] = [];

  EVENTS.forEach(([title, hour, description], eventNumber) => {
    const landmark = photogenic[eventNumber];
    if (!landmark) return;

    const metadata = buildMetadata(landmark, fetchedAt, busynessByPlace.get(landmark.place_id)!);
    metadata.kind = 'event';
    metadata.event_time = eventTimeUtc(fetchedAt, hour);
    metadata.heat_score = 6;

    docs.push({
      text: `Event: ${title}. ${description.replace('{name}', landmark.name)}`,
      metadata,
    });
  });

  return docs;
}

function buildCaptionDocs(
  photogenic: Landmark[],
  fetchedAt: string,
  busynessByPlace: Map<string, number>,
): Document[] {
  if (!photogenic.length) return [];

  return CAPTIONS.map((caption, captionNumber) => {
    const landmark = photogenic[captionNumber % photogenic.length];
    const metadata = buildMetadata(landmark, fetchedAt, busynessByPlace.get(landmark.place_id)!);
    metadata.kind = 'caption';
    metadata.heat_score = (captionNumber % 10) + 1;

    return {
      text: `Instagram-style caption about ${landmark.name}: ${caption}`,
      metadata,
    };
  });
}

/**
 * The local calendar date in San Francisco at `fetchedAt`, combined with `hour`
 * local time, expressed back in UTC.
 */
function eventTimeUtc(fetchedAt: string, hour: number): string {
  const offsetMs = SAN_FRANCISCO_UTC_OFFSET_HOURS * 60 * 60 * 1000;

  // Shifting into local time lets the UTC getters read off local calendar
  // fields, which is how the tonight-in-SF date is derived without tzdata.
  const local = new Date(new Date(fetchedAt).getTime() + offsetMs);
  const localMidnightUtc = Date.UTC(
    local.getUTCFullYear(),
    local.getUTCMonth(),
    local.getUTCDate(),
  );

  return asUtc(new Date(localMidnightUtc + hour * 60 * 60 * 1000 - offsetMs));
}

function buildMetadata(
  landmark: Landmark,
  fetchedAt: string,
  busynessPct: number,
): Record<string, unknown> {
  return {
    ...landmark,
    busyness_pct: busynessPct,
    busyness_at: fetchedAt,
    heat_score: 1,
    open_now: true,
    event_time: null,
    lang: 'en',
    source: 'fixture',
    fetched_at: fetchedAt,
    is_fixture: true,
  };
}

function asUtc(value: Date): string {
  return value.toISOString().replace(/\.\d{3}Z$/, 'Z');
}
